engine/context/src/main/java/io/deephaven/engine/context/QueryCompiler.java

```diff
@@ -144,327 +193,255 @@ public static void writeClass(final File destinationDirectory, final String clas
-            // that does not have sufficient security permissions.
-            this.ucl = doPrivileged((PrivilegedAction<WritableURLClassLoader>) () -> new WritableURLClassLoader(urls,
-                    parentClassLoader));
-
-            if (isCacheDirectory) {
-                addClassSource(classDestination);
+                return promise.get();
+            } catch (InterruptedException | ExecutionException error) {
+                throw new UncheckedDeephavenException(error);
             }
         }
 
-        @Override
-        public Map<String, CompletableFuture<Class<?>>> getKnownClasses() {
-            return knownClasses;
+        // It's my job to fulfill the promise
+        try {
+            return compileHelper(className, classBody, packageNameRoot, codeLog, parameterClasses);
+        } catch (RuntimeException e) {
+            promise.completeExceptionally(e);
+            throw e;
         }
+    }
 
-        @Override
-        public ClassLoader getClassLoaderForFormula(final Map<String, Class<?>> parameterClasses) {
-            // We should always be able to get our own class loader, even if this is invoked from external code
-            // that doesn't have security permissions to make ITS own class loader.
-            return doPrivileged((PrivilegedAction<URLClassLoader>) () -> new URLClassLoader(ucl.getURLs(), ucl) {
-                // Once we find a class that is missing, we should not attempt to load it again,
-                // otherwise we can end up with a StackOverflow Exception
-                final HashSet<String> missingClasses = new HashSet<>();
-
-                @Override
-                protected Class<?> findClass(String name) throws ClassNotFoundException {
-                    // If we have a parameter that uses this class, return it
-                    final Class<?> paramClass = parameterClasses.get(name);
-                    if (paramClass != null) {
-                        return paramClass;
-                    }
-
-                    // Unless we are looking for a formula or Groovy class, we should use the default behavior
-                    if (!isFormulaClass(name)) {
-                        return super.findClass(name);
-                    }
-
-                    // if it is a groovy class, always try to use the instance in the shell
-                    if (name.startsWith(DYNAMIC_GROOVY_CLASS_PREFIX)) {
-                        try {
-                            return ucl.getParent().loadClass(name);
-                        } catch (final ClassNotFoundException ignored) {
-                            // we'll try to load it otherwise
-                        }
-                    }
+    private static void ensureDirectories(final File file, final Supplier<String> runtimeErrMsg) {
+        // File.mkdirs() checks for existrance on entry, in which case it returns false.
```

Comment:
Can you correct this old typo?
```suggestion
        // File.mkdirs() checks for existence on entry, in which case it returns false.
```